### PR TITLE
Account locking for JIT provisioned users

### PR DESCRIPTION
**Pull request by GANGANI**
### Description

This PR changes will enable account locking capability for the JIT provisioned users.

Related issue: https://github.com/wso2-enterprise/asgardeo-product/issues/4651

### CODE REVIEW

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/JITProvisioningPostAuthenticationHandler.java [hunk 1]
```diff
@@ -354,6 +366,44 @@ private boolean isUserNameFoundFromUserIDClaimURI(Map<String, String> localClaim
...
+
+    private void handleAccountLockLoginFailure(String retryPage, AuthenticationContext context,
+                                               HttpServletResponse response) throws PostAuthenticationFailedException {
+
+        try {
+            // ToDo: Add support to configure enable/disable authentication failure reason.
+            boolean showAuthFailureReason = true;
+            StringBuilder urlBuilder = new StringBuilder();
+            urlBuilder.append(retryPage);
+            urlBuilder.append("?sp=").append(context.getServiceProviderName());
```
Comment (emswbandara) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/JITProvisioningPostAuthenticationHandler.java:390]: better to use FrameworkUtils.appendQueryParams for this

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/JITProvisioningPostAuthenticationHandler.java [hunk 2]
```diff
@@ -334,6 +338,14 @@ private PostAuthnHandlerFlowStatus handleRequestFlow(HttpServletRequest request,
                         username = sequenceConfig.getAuthenticatedUser().getUserName();
                         isUserCreated = true;
                     }
+                    // Check if the associated local account is locked.
+                    if (isAccountLocked(username, context.getTenantDomain())) {
+                        if (log.isDebugEnabled()) {
+                            log.debug("The account is locked for the user " + username);
```
Comment (somindatommy) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/JITProvisioningPostAuthenticationHandler.java:344]: Shall we mention the tenant domain as well? WDYT?

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/internal/FrameworkServiceDataHolder.java [hunk 3]
```diff
@@ -529,4 +531,14 @@ public void setServerSessionManagementService(
...
     }
+
+    public void setAccountLockService(AccountLockService accountLockService) {
+
+        this.accountLockService = accountLockService;
+    }
+
+    public AccountLockService getAccountLockService() {
+
+        return accountLockService;
```
Comment (somindatommy) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/internal/FrameworkServiceDataHolder.java:542]: Lets validate whether the `accountLockService` is not or not. If null, lets throw an runtime exception

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/JITProvisioningPostAuthenticationHandler.java (hunk 1 above)
Comment (GANGANI) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/JITProvisioningPostAuthenticationHandler.java]: Fixed https://github.com/wso2/carbon-identity-framework/pull/3598/commits/44b14a3dc4c0be84122db9c92c09d08ce2daf877

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/JITProvisioningPostAuthenticationHandler.java (hunk 2 above)
Comment (GANGANI) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/JITProvisioningPostAuthenticationHandler.java]: Fixed https://github.com/wso2/carbon-identity-framework/pull/3598/commits/44b14a3dc4c0be84122db9c92c09d08ce2daf877

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/internal/FrameworkServiceDataHolder.java (hunk 3 above)
Comment (mevan-karu) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/internal/FrameworkServiceDataHolder.java:542]: In here https://github.com/wso2/carbon-identity-framework/pull/3598/files#diff-8bccdb72ac4633653cb8821508b1c64219f85e428f8bf61fe51c7540c8946b4fR793 this service is marked as mandatory. Service won't be activated if AccountLockService is not set. Why do we need to throw a runtime exception?

Comment (somindatommy) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/internal/FrameworkServiceDataHolder.java:542]: Agree, I have seen in some places where we have done that as a practice feel to ack that

**Review by mevan-karu — APPROVED** [review]

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/JITProvisioningPostAuthenticationHandler.java [hunk 4]
```diff
@@ -334,6 +338,15 @@ private PostAuthnHandlerFlowStatus handleRequestFlow(HttpServletRequest request,
                         username = sequenceConfig.getAuthenticatedUser().getUserName();
                         isUserCreated = true;
                     }
+                    // Check if the associated local account is locked.
```
Comment (IsuraD) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/JITProvisioningPostAuthenticationHandler.java:341]: how do we lock the user if the user is trying continuos invalid codes?

Comment (GANGANI) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/JITProvisioningPostAuthenticationHandler.java:341]: It is handled by the fix: https://github.com/wso2-extensions/identity-outbound-auth-totp/pull/73/files

**Review by ashensw — APPROVED** [review]

**Pull request merged**